python/GafferSceneTest/ShaderTest.py

```diff
@@ -361,5 +361,79 @@ def testNameSwitch( self ) :
 			)
 
+	def testSpline( self ) :
+
+		n1 = GafferSceneTest.TestShader( "n1" )
+
+		n2 = GafferSceneTest.TestShader( "n2" )
+		n2["type"].setValue( "test:surface" )
+
+		n2["parameters"]["i"].setInput( n1["out"]["r"] )
+
+		network = n2.attributes()["test:surface"]
+
+		self.assertEqual( network.shaders()["n1"].parameters["spline"], IECore.SplinefColor3fData() )
+
+		n1["parameters"]["spline"].addPoint()
+		n1["parameters"]["spline"].addPoint()
+		n1["parameters"]["spline"].addPoint()
+		n1["parameters"]["spline"]["p1"]["x"].setValue( 1 )
+		n1["parameters"]["spline"]["p1"]["y"].setValue( imath.Color3f( 1 ) )
+		n1["parameters"]["spline"]["p2"]["x"].setValue( 0.6 )
+		n1["parameters"]["spline"]["p2"]["y"].setValue( imath.Color3f( 0.4, 0.5, 0.7 ) )
+
+
+		network = n2.attributes()["test:surface"]
+
+		refSpline = IECore.SplinefColor3f()
+		refSpline[0] = imath.Color3f( 0 )
+		refSpline[0] = imath.Color3f( 0 )
+		refSpline[0.6] = imath.Color3f( 0.4, 0.5, 0.7 )
+		refSpline[1] = imath.Color3f( 1 )
+		refSpline[1] = imath.Color3f( 1 )
+
+		self.assertEqual( network.shaders()["n1"].parameters["spline"].value, refSpline )
+
+		inN1 = GafferSceneTest.TestShader( "inN1" )
+
+		n1["parameters"]["spline"]["p0"]["y"].setInput( inN1["out"] )
+		n1["parameters"]["spline"]["p1"]["y"]["b"].setInput( inN1["out"]["g"] )
+
+		network = n2.attributes()["test:surface"]
+
+		self.assertEqual( len( network ), 3 )
+		self.assertEqual(
+			network.inputConnections( "n1" ), [
+				network.Connection( network.Parameter( "inN1", "", ), network.Parameter( "n1", "spline[0].y" ) ),
+				network.Connection( network.Parameter( "inN1", "", ), network.Parameter( "n1", "spline[1].y" ) ),
+				network.Connection( network.Parameter( "inN1", "g", ), network.Parameter( "n1", "spline[3].y.b" ) ),
+				network.Connection( network.Parameter( "inN1", "g", ), network.Parameter( "n1", "spline[4].y.b" ) )
+			]
+		)
+
+		n1["parameters"]["spline"]["p1"]["x"].setInput( inN1["out"]["g"] )
+
+		with self.assertRaisesRegex( RuntimeError, "n2.__outAttributes : Shader connections to n1.parameters.spline.p1.x are not supported." ) :
+			network = n2.attributes()["test:surface"]
+
+		n1["parameters"]["spline"]["p1"]["x"].setInput( None )
+
+		n1["parameters"]["spline"]["interpolation"].setValue( Gaffer.SplineDefinitionInterpolation.Linear )
+
+		network = n2.attributes()["test:surface"]
+		self.assertEqual(
+			network.inputConnections( "n1" ), [
+				network.Connection( network.Parameter( "inN1", "", ), network.Parameter( "n1", "spline[0].y" ) ),
+				network.Connection( network.Parameter( "inN1", "g", ), network.Parameter( "n1", "spline[2].y.b" ) )
+			]
+		)
+
+		n1["parameters"]["spline"]["interpolation"].setValue( Gaffer.SplineDefinitionInterpolation.MonotoneCubic )
+
+		with self.assertRaisesRegex( RuntimeError, "n2.__outAttributes : Cannot support monotone cubic interpolation for splines with inputs, for plug n1.parameters.spline" ):
```

Comment:
What would be required to be able to support this? Moving the monotone-cubic logic inside the shader, where the `y` values are available? Definitely not suggesting we should do anything now - I'm just curious.

Reply:
Correct, we would need to move the tangent calculation logic inside the shader, and then reproduce that logic for drawing the preview in the spline widget.